### ISSUE TITLE
Supporting SEV-SNP guests on KVM

### ISSUE THREAD

**Issue by ymjing** [cloud-hypervisor/cloud-hypervisor#6653]
Hi CloudHypervisor team,

I'd like to create an issue to track the work of supporting SNP guests on KVM. 

The idea was discussed with @likebreath weeks ago; overall we'd like to contribute to CloudHypervisor so that it can boot [Oak](https://github.com/project-oak/oak) Stage0, Oak Restricted Kernel, and Oak Containers in SEV-SNP guests on KVM.

The work will be based on Linux kernel 6.11, which contains the SNP specific KVM APIs upstreamed by AMD. Some high-level tasks are:

- [ ] `sev_snp` feature - decouple from`igvm` and `mshv`; @jinankjain
- [ ]  `hypervisor/kvm` changes - in particular, implement `sev_snp_init`, `import_isolated_pages`, and `complete_isolated_import` with `KVM_SEV_SNP_*` ioctls;
- [ ] `vmm` changes: load firmware, set up CPUID & SECRETS pages, initialize AP registers.
- [ ] Replacement of`fw_cfg` which is not available in CHV but needed for OVMF/Oak Stage0 to receive kernel and initrd information.

Please let me know what you think.

**Comment by liuw** [comment]
About that last point on `fw_cfg`, what's the replacement you have in mind?

**Comment by likebreath** [comment]
@ymjing Welcome to the community. I am glad to see your team is expanding Cloud Hypervisor's capability to support your use cases. Look forward to the contributions from your team.

As for contacts in the community, @liuw is the best person to talk with regarding the support of mshv and sev-snp in general. Also, you may already know, @russell-islam, @jinankjain, and @NunoDasNeves have also been contributing to the related features. Of course, @rbradford and myself are here to help when needed, say discussing large refactoring, providing feedbacks, and reviewing PRs in general. 

The steps you laid-out look good to me. One comment on the kernel 6.11 you mentioned.  Our code base right now relies on kernel 6.2 for KVM APIs and data structures, through the use of kvm-bindings and kvm-ioctls crate. You will need to carry any new KVM APIs or data structures from v6.11 in your own patches with appropriate feature gates. This is the practice we have been using, say for TDX support that needs newer Linux kernel. (A side note, the latest kvm-bindings crate release v0.9.0 moved to kernel v6.9 [1], and we will upgrade to that in the near future.)

[1] https://github.com/rust-vmm/kvm-bindings/releases/tag/v0.9.0

**Comment by ymjing** [comment]
> About that last point on `fw_cfg`, what's the replacement you have in mind?
> 
> What format do you have in mind for packaging the guest image? Do you plan to use an SVSM?
> 
> ~What firmware do you have in mind? OVMF?~

tbh I don't have a good one in my mind. SMBIOS, e820, ACPI, HOB, to name a few. Whichever we pick would have to be implemented in both CHV and the firmware. Not sure if OVMF is OK with such changes, though.

Our choice of firmware is [Oak Stage0](https://github.com/project-oak/oak/tree/main/stage0). OVMF can be a stretch goal; but optional.

I'd prefer a general packaging format recommended by the CHV community. So far CHV uses IGVM with MSHV. I'd like to know what you think - should we use IGVM too for KVM?

**Comment by ymjing** [comment]
> @ymjing Welcome to the community. I am glad to see your team is expanding Cloud Hypervisor's capability to support your use cases. Look forward to the contributions from your team.
> 
> As for contacts in the community, @liuw is the best person to talk with regarding the support of mshv and sev-snp in general. Also, you may already know, @russell-islam, @jinankjain, and @NunoDasNeves have also been contributing to the related features. Of course, @rbradford and myself are here to help when needed, say discussing large refactoring, providing feedbacks, and reviewing PRs in general.
> 
> The steps you laid-out look good to me. One comment on the kernel 6.11 you mentioned. Our code base right now relies on kernel 6.2 for KVM APIs and data structures, through the use of kvm-bindings and kvm-ioctls crate. You will need to carry any new KVM APIs or data structures from v6.11 in your own patches with appropriate feature gates. This is the practice we have been using, say for TDX support that needs newer Linux kernel. (A side note, the latest kvm-bindings crate release v0.9.0 moved to kernel v6.9 [1], and we will upgrade to that in the near future.)
> 
> [1] https://github.com/rust-vmm/kvm-bindings/releases/tag/v0.9.0

Thanks Bo! So far I'm able to use `kvm-ioctls` 0.18.0  and `kvm-bindings` 0.9.0 on Linux kernel 6.11 rc2. Please refer to my small PoC here: https://github.com/ymjing/mini-snp.

**Comment by likebreath** [comment]
> Thanks Bo! So far I'm able to use `kvm-ioctls` 0.18.0 and `kvm-bindings` 0.9.0 on Linux kernel 6.11 rc2. Please refer to my small PoC here: https://github.com/ymjing/mini-snp.

That's great to hear. I will bump our dependencies on kvm-ioctl and kvm-bindings soon, which requires the vfio crates to be bumped first [1].

[1] https://github.com/rust-vmm/vfio/pull/57

**Comment by liuw** [comment]
> > Thanks Bo! So far I'm able to use `kvm-ioctls` 0.18.0 and `kvm-bindings` 0.9.0 on Linux kernel 6.11 rc2. Please refer to my small PoC here: https://github.com/ymjing/mini-snp.
> 
> That's great to hear. I will bump our dependencies on kvm-ioctl and kvm-bindings soon, which requires the vfio crates to be bumped first [1].
> 
> [1] [rust-vmm/vfio#57](https://github.com/rust-vmm/vfio/pull/57)

The vfio crate has been updated to the latest dependencies.

**Comment by likebreath** [comment]
> The vfio crate has been updated to the latest dependencies.

Great. Just sent the PR for bumping our dependencies #6659.

**Comment by jinankjain** [comment]
I think IGVM would be a good choice for packaging the guest kernel image. What is the file format of Oak Stage0 firmware. We currently support bzImage and ELF binaries. If this is also an ELF binary then we can probably re-use the existing IGVM infrastructure.

**Comment by dingelish** [comment]
@jinankjain Stage0 firmware is a regular 2MB binary. Hypervisor usually maps the firmware to the top 2MB space. Stage0 currently only supports direct kernel boot. On QEMU we use `-kernel` and `-initrd` for that. Vanilla Linux kernel works.

**Comment by ymjing** [comment]
> I think IGVM would be a good choice for packaging the guest kernel image. What is the file format of Oak Stage0 firmware. We currently support bzImage and ELF binaries. If this is also an ELF binary then we can probably re-use the existing IGVM infrastructure.

Yeah I was able to pack Stage0 into an IGVM file. The current IGVM loader has no problem loading the Data, CPUID, and SECRET pages into guest memory. However, CHV cannot load the VGSA page which is supposed to be at 0xffff_ffff_f000. 

@jinankjain how does CHV's IGVM loader set the intiial vCPU registers?

**Comment by jinankjain** [comment]
We are loading the initial vCPU registers via VMSA page only. If you check the igvm_loader code in CH we are passing the GPA of VMSA via the entry point. https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/vmm/src/vm.rs#L1031

**Comment by liuw** [comment]
@ymjing Hey, Yiming, I'm wondering how this project is going?

**Comment by ymjing** [comment]
Hi, the team has been working on an internal fork of CloudHypervisor. It's mostly working as of now with the following features: (1) implementation SNP_LAUNCH_START/UPDATE/FINISH; (2) Oak Stage0 loader with the IGVM loader, (3) GHCB support, and (4) a basic implementation of `fw_cfg`. 

We'd like to polish these features a bit longer before we share with the CHV community here.

**Comment by likebreath** [comment]
The initial RFC posted by @AlexOrozco1256 in #6993:

> We are working with the project [Oak](https://github.com/project-oak/oak) team to enable running [Oak Containers](https://github.com/project-oak/oak/tree/main/oak_containers) on cloud hypervisor. These containers are SEV-SNP enlightened and can be used alongside with the rest of the oak suite for attestation and measurement of the cvm.
> 
> Here is the functionality we are adding to CHV.
> 
> 1. Implmenting the sev-snp APIs previously defined (sev_snp_init, import_isolated_pages, complete_isolated_import).
> 2. Expanded the igvm loader to support booting KVM + SEV-SNP enabled VMs. (Note that we can also use the igvm loader to boot VMs with standard cloud images with/without sev-snp enabled as seen here (https://github.com/AlexOrozco1256/chv-demo)
> 3. Added a fw_cfg device with DMA enabled. Oak's [Stage0](https://github.com/project-oak/oak/tree/main/stage0_bin) firmware uses fw_cfg for loading various parameters into the VM.
> 
> Additional design/implementation details can be found here: https://tinyurl.com/chv-kvm-sev-snp

Let's use this issue to track the progress of SEV-SNP support on KVM, while using the RFC #6993 or the [google document above ](https://tinyurl.com/chv-kvm-sev-snp) for commenting specific questions wherever you find suitable. 

\cc @ymjing @rbradford @liuw @jinankjain @russell-islam
